{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.11.13","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"},"kaggle":{"accelerator":"none","dataSources":[{"sourceId":116062,"databundleVersionId":14084779,"sourceType":"competition"},{"sourceId":8236972,"sourceType":"datasetVersion","datasetId":4885707},{"sourceId":13588728,"sourceType":"datasetVersion","datasetId":8627154},{"sourceId":13920669,"sourceType":"datasetVersion","datasetId":8637089}],"dockerImageVersionId":31153,"isInternetEnabled":true,"language":"python","sourceType":"notebook","isGpuEnabled":false}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Install required libraries","metadata":{}},{"cell_type":"code","source":"!pip install obonet networkx","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:10:47.625854Z","iopub.execute_input":"2025-12-04T04:10:47.626170Z","iopub.status.idle":"2025-12-04T04:10:53.716924Z","shell.execute_reply.started":"2025-12-04T04:10:47.626144Z","shell.execute_reply":"2025-12-04T04:10:53.715681Z"}},"outputs":[{"name":"stdout","text":"Collecting obonet\n  Downloading obonet-1.1.1-py3-none-any.whl.metadata (6.7 kB)\nRequirement already satisfied: networkx in /usr/local/lib/python3.11/dist-packages (3.5)\nDownloading obonet-1.1.1-py3-none-any.whl (9.2 kB)\nInstalling collected packages: obonet\nSuccessfully installed obonet-1.1.1\n","output_type":"stream"}],"execution_count":1},{"cell_type":"code","source":"!pip install duckdb --no-index --find-links=/kaggle/input/polars-and-duckdb/kaggle/working/mysitepackages/duck_pkg\n!pip install polars --no-index --find-links=/kaggle/input/polars-and-duckdb/kaggle/working/mysitepackages/polars_pkg","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:01.453213Z","iopub.execute_input":"2025-12-04T04:11:01.453608Z","iopub.status.idle":"2025-12-04T04:11:09.168711Z","shell.execute_reply.started":"2025-12-04T04:11:01.453573Z","shell.execute_reply":"2025-12-04T04:11:09.167520Z"}},"outputs":[{"name":"stdout","text":"Looking in links: /kaggle/input/polars-and-duckdb/kaggle/working/mysitepackages/duck_pkg\nRequirement already satisfied: duckdb in /usr/local/lib/python3.11/dist-packages (1.3.2)\nLooking in links: /kaggle/input/polars-and-duckdb/kaggle/working/mysitepackages/polars_pkg\nRequirement already satisfied: polars in /usr/local/lib/python3.11/dist-packages (1.25.0)\n","output_type":"stream"}],"execution_count":2},{"cell_type":"code","source":"!pip install biopython","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:14.648617Z","iopub.execute_input":"2025-12-04T04:11:14.648972Z","iopub.status.idle":"2025-12-04T04:11:20.275179Z","shell.execute_reply.started":"2025-12-04T04:11:14.648943Z","shell.execute_reply":"2025-12-04T04:11:20.274167Z"}},"outputs":[{"name":"stdout","text":"Collecting biopython\n  Downloading biopython-1.86-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\nRequirement already satisfied: numpy in /usr/local/lib/python3.11/dist-packages (from biopython) (1.26.4)\nRequirement already satisfied: mkl_fft in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.3.8)\nRequirement already satisfied: mkl_random in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (1.2.4)\nRequirement already satisfied: mkl_umath in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (0.1.1)\nRequirement already satisfied: mkl in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2025.2.0)\nRequirement already satisfied: tbb4py in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2022.2.0)\nRequirement already satisfied: mkl-service in /usr/local/lib/python3.11/dist-packages (from numpy->biopython) (2.4.1)\nRequirement already satisfied: intel-openmp<2026,>=2024 in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2024.2.0)\nRequirement already satisfied: tbb==2022.* in /usr/local/lib/python3.11/dist-packages (from mkl->numpy->biopython) (2022.2.0)\nRequirement already satisfied: tcmlib==1.* in /usr/local/lib/python3.11/dist-packages (from tbb==2022.*->mkl->numpy->biopython) (1.4.0)\nRequirement already satisfied: intel-cmplr-lib-rt in /usr/local/lib/python3.11/dist-packages (from mkl_umath->numpy->biopython) (2024.2.0)\nRequirement already satisfied: intel-cmplr-lib-ur==2024.2.0 in /usr/local/lib/python3.11/dist-packages (from intel-openmp<2026,>=2024->mkl->numpy->biopython) (2024.2.0)\nDownloading biopython-1.86-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m28.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n\u001b[?25hInstalling collected packages: biopython\nSuccessfully installed biopython-1.86\n","output_type":"stream"}],"execution_count":3},{"cell_type":"markdown","source":"## Import required modules","metadata":{}},{"cell_type":"code","source":"import numpy as np\nimport pandas as pd\nimport os\nimport duckdb as dd\nimport polars as pl\nimport h5py\nimport networkx as nx\nimport obonet as ob\n\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\nfrom Bio import SeqIO","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:24.666361Z","iopub.execute_input":"2025-12-04T04:11:24.666703Z","iopub.status.idle":"2025-12-04T04:11:27.585116Z","shell.execute_reply.started":"2025-12-04T04:11:24.666668Z","shell.execute_reply":"2025-12-04T04:11:27.584143Z"}},"outputs":[],"execution_count":4},{"cell_type":"markdown","source":"## Separating out the GO terms based on Ontology roots","metadata":{}},{"cell_type":"code","source":"def separate_go_terms_by_namespace(graph):\n    \"\"\"\n    Separates GO terms from a networkx graph into BP, CC, and MF categories.\n\n    Args:\n        graph: A networkx MultiDiGraph loaded with obonet.\n\n    Returns:\n        A dictionary containing lists of GO term IDs for each namespace.\n    \"\"\"\n    bp_terms = []\n    cc_terms = []\n    mf_terms = []\n\n    # Iterate through every node in the graph, accessing its ID (node_id) and attributes (data)\n    for node_id, data in graph.nodes(data=True):\n        # The namespace is stored in the 'namespace' key of the node's data dictionary\n        namespace = data.get('namespace')\n\n        if namespace == 'biological_process':\n            bp_terms.append(node_id)\n        elif namespace == 'cellular_component':\n            cc_terms.append(node_id)\n        elif namespace == 'molecular_function':\n            mf_terms.append(node_id)\n\n    return {\n        'BP': bp_terms,\n        'CC': cc_terms,\n        'MF': mf_terms\n    }","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:34.055091Z","iopub.execute_input":"2025-12-04T04:11:34.055713Z","iopub.status.idle":"2025-12-04T04:11:34.062491Z","shell.execute_reply.started":"2025-12-04T04:11:34.055682Z","shell.execute_reply":"2025-12-04T04:11:34.061428Z"}},"outputs":[],"execution_count":5},{"cell_type":"markdown","source":"## Load the Gene Ontology (GO) basic OBO file from the OBO Library","metadata":{}},{"cell_type":"code","source":"go_file_path = '/kaggle/input/cafa-6-protein-function-prediction/Train/go-basic.obo'\nprint(f\"Loading Gene Ontology from {go_file_path}...\")\ngo_graph = ob.read_obo(go_file_path)\nprint(f\"Total terms loaded: {len(go_graph)}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:40.525151Z","iopub.execute_input":"2025-12-04T04:11:40.525573Z","iopub.status.idle":"2025-12-04T04:11:47.452974Z","shell.execute_reply.started":"2025-12-04T04:11:40.525511Z","shell.execute_reply":"2025-12-04T04:11:47.452052Z"}},"outputs":[{"name":"stdout","text":"Loading Gene Ontology from /kaggle/input/cafa-6-protein-function-prediction/Train/go-basic.obo...\nTotal terms loaded: 40122\n","output_type":"stream"}],"execution_count":6},{"cell_type":"markdown","source":"## Separate the terms","metadata":{}},{"cell_type":"code","source":"separated_terms = separate_go_terms_by_namespace(go_graph)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:49.560805Z","iopub.execute_input":"2025-12-04T04:11:49.561096Z","iopub.status.idle":"2025-12-04T04:11:49.578438Z","shell.execute_reply.started":"2025-12-04T04:11:49.561075Z","shell.execute_reply":"2025-12-04T04:11:49.577391Z"}},"outputs":[],"execution_count":7},{"cell_type":"code","source":"print(\"\\nGO Term Counts by Namespace:\")\nprint(f\"Biological Process (BP): {len(separated_terms['BP'])} terms\")\nprint(f\"Cellular Component (CC): {len(separated_terms['CC'])} terms\")\nprint(f\"Molecular Function (MF): {len(separated_terms['MF'])} terms\")\n\nprint(\"\\nFirst 5 Cellular Component (CC) terms:\")\nfor term_id in separated_terms['CC'][:5]:\n    # You can get the human-readable name using the graph data dictionary\n    name = go_graph.nodes[term_id].get('name')\n    print(f\"- {term_id}: {name}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:11:53.455676Z","iopub.execute_input":"2025-12-04T04:11:53.455986Z","iopub.status.idle":"2025-12-04T04:11:53.462648Z","shell.execute_reply.started":"2025-12-04T04:11:53.455962Z","shell.execute_reply":"2025-12-04T04:11:53.461569Z"}},"outputs":[{"name":"stdout","text":"\nGO Term Counts by Namespace:\nBiological Process (BP): 25950 terms\nCellular Component (CC): 4041 terms\nMolecular Function (MF): 10131 terms\n\nFirst 5 Cellular Component (CC) terms:\n- GO:0000015: phosphopyruvate hydratase complex\n- GO:0000109: nucleotide-excision repair complex\n- GO:0000110: nucleotide-excision repair factor 1 complex\n- GO:0000111: nucleotide-excision repair factor 2 complex\n- GO:0000112: nucleotide-excision repair factor 3 complex\n","output_type":"stream"}],"execution_count":8},{"cell_type":"code","source":"go_graph_edges = list(go_graph.edges)\n\nprint(go_graph_edges[0])\nprint(go_graph.nodes[\"GO:0008150\"])\nprint(\"*************\")\nprint(go_graph.nodes[\"GO:0005575\"])\nprint(\"*************\")\nprint(go_graph.nodes[\"GO:0003674\"])","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:23:41.482390Z","iopub.execute_input":"2025-12-04T04:23:41.482760Z","iopub.status.idle":"2025-12-04T04:23:41.552701Z","shell.execute_reply.started":"2025-12-04T04:23:41.482732Z","shell.execute_reply":"2025-12-04T04:23:41.551608Z"}},"outputs":[{"name":"stdout","text":"('GO:0000001', 'GO:0048308', 'is_a')\n{'name': 'biological_process', 'namespace': 'biological_process', 'alt_id': ['GO:0000004', 'GO:0007582', 'GO:0044699'], 'def': '\"A biological process is the execution of a genetically-encoded biological module or program. It consists of all the steps required to achieve the specific biological objective of the module. A biological process is accomplished by a particular set of molecular functions carried out by specific gene products (or macromolecular complexes), often in a highly regulated manner and in a particular temporal sequence.\" [GOC:pdt]', 'comment': \"Note that, in addition to forming the root of the biological process ontology, this term is recommended for the annotation of gene products whose biological process is unknown. When this term is used for annotation, it indicates that no information was available about the biological process of the gene product annotated as of the date the annotation was made; the evidence code 'no data' (ND), is used to indicate this.\", 'subset': ['goslim_candida', 'goslim_chembl', 'goslim_metagenomics', 'goslim_pir', 'goslim_plant', 'goslim_pombe', 'goslim_yeast'], 'synonym': ['\"biological process\" EXACT []', '\"physiological process\" EXACT []', '\"single organism process\" RELATED []', '\"single-organism process\" RELATED []'], 'xref': ['Wikipedia:Biological_process']}\n*************\n{'name': 'cellular_component', 'namespace': 'cellular_component', 'alt_id': ['GO:0008372'], 'def': '\"A location, relative to cellular compartments and structures, occupied by a macromolecular machine. There are three types of cellular components described in the gene ontology: (1) the cellular anatomical entity where a gene product carries out a molecular function (e.g., plasma membrane, cytoskeleton) or membrane-enclosed compartments (e.g., mitochondrion); (2) virion components, where viral proteins act, and (3) the stable macromolecular complexes of which gene product are parts (e.g., the clathrin complex).\" [GOC:pdt]', 'comment': \"Note that, in addition to forming the root of the cellular component ontology, this term is recommended for the annotation of gene products whose cellular component is unknown. When this term is used for annotation, it indicates that no information was available about the cellular component of the gene product annotated as of the date the annotation was made; the evidence code 'no data' (ND), is used to indicate this.\", 'subset': ['goslim_candida', 'goslim_chembl', 'goslim_metagenomics', 'goslim_pir', 'goslim_plant', 'goslim_yeast'], 'synonym': ['\"cell or subcellular entity\" EXACT []', '\"cellular component\" EXACT []', '\"subcellular entity\" RELATED [NIF_Subcellular:nlx_subcell_100315]'], 'xref': ['NIF_Subcellular:sao1337158144']}\n*************\n{'name': 'molecular_function', 'namespace': 'molecular_function', 'alt_id': ['GO:0005554'], 'def': '\"A molecular process that can be carried out by the action of a single macromolecular machine, usually via direct physical interactions with other molecular entities. Function in this sense denotes an action, or activity, that a gene product (or a complex) performs.\" [GOC:pdt]', 'comment': \"Note that, in addition to forming the root of the molecular function ontology, this term is recommended for the annotation of gene products whose molecular function is unknown. When this term is used for annotation, it indicates that no information was available about the molecular function of the gene product annotated as of the date the annotation was made; the evidence code 'no data' (ND), is used to indicate this. Despite its name, this is not a type of 'function' in the sense typically defined by upper ontologies such as Basic Formal Ontology (BFO). It is instead a BFO:process carried out by a single gene product or complex.\", 'subset': ['goslim_candida', 'goslim_chembl', 'goslim_metagenomics', 'goslim_pir', 'goslim_plant', 'goslim_yeast'], 'synonym': ['\"molecular function\" EXACT []']}\n","output_type":"stream"}],"execution_count":18},{"cell_type":"code","source":"id_to_name = {id_: data.get(\"name\") for id_, data in go_graph.nodes(data=True)}\n\nname_to_id = {\n    data[\"name\"]: id_ for id_, data in go_graph.nodes(data=True) if \"name\" in data\n}\n\nprint(\"id_to_name['GO:0008150'] -> \", id_to_name['GO:0008150'])\nprint(\"name_to_id['cellular_component'] -> \", name_to_id['cellular_component'])","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:47:33.131858Z","iopub.execute_input":"2025-12-04T04:47:33.132126Z","iopub.status.idle":"2025-12-04T04:47:33.182129Z","shell.execute_reply.started":"2025-12-04T04:47:33.132108Z","shell.execute_reply":"2025-12-04T04:47:33.181238Z"}},"outputs":[{"name":"stdout","text":"id_to_name['GO:0008150'] ->  biological_process\nname_to_id['cellular_component'] ->  GO:0005575\n","output_type":"stream"}],"execution_count":37},{"cell_type":"code","source":"# Find edges to parent terms\nnode = name_to_id[\"cellular_component\"]\nfor child, parent, key in go_graph.out_edges(node, keys=True):\n    print(f\"• {id_to_name[child]} ⟶ {key} ⟶ {id_to_name[parent]}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T04:47:37.495358Z","iopub.execute_input":"2025-12-04T04:47:37.496256Z","iopub.status.idle":"2025-12-04T04:47:37.500768Z","shell.execute_reply.started":"2025-12-04T04:47:37.496225Z","shell.execute_reply":"2025-12-04T04:47:37.499865Z"}},"outputs":[],"execution_count":38},{"cell_type":"code","source":"# Find edges to children terms\nnode = name_to_id[\"cellular_component\"]\nfor child, parent, key in go_graph.in_edges(node, keys=True):\n    print(f\"• {id_to_name[parent]} ⟵ {key} ⟵ {id_to_name[child]}\")\nprint(\"**********************\")\nnode = name_to_id[\"biological_process\"]\nfor child, parent, key in go_graph.in_edges(node, keys=True):\n    print(f\"• {id_to_name[parent]} ⟵ {key} ⟵ {id_to_name[child]}\")\nprint(\"**********************\")\nnode = name_to_id[\"molecular_function\"]\nfor child, parent, key in go_graph.in_edges(node, keys=True):\n    print(f\"• {id_to_name[parent]} ⟵ {key} ⟵ {id_to_name[child]}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T05:05:37.253863Z","iopub.execute_input":"2025-12-04T05:05:37.254187Z","iopub.status.idle":"2025-12-04T05:05:37.261842Z","shell.execute_reply.started":"2025-12-04T05:05:37.254162Z","shell.execute_reply":"2025-12-04T05:05:37.260864Z"}},"outputs":[{"name":"stdout","text":"• cellular_component ⟵ is_a ⟵ protein-containing complex\n• cellular_component ⟵ is_a ⟵ virion component\n• cellular_component ⟵ is_a ⟵ cellular anatomical structure\n**********************\n• biological_process ⟵ is_a ⟵ immune system process\n• biological_process ⟵ is_a ⟵ cellular process\n• biological_process ⟵ is_a ⟵ viral process\n• biological_process ⟵ is_a ⟵ reproductive process\n• biological_process ⟵ is_a ⟵ multicellular organismal process\n• biological_process ⟵ is_a ⟵ developmental process\n• biological_process ⟵ is_a ⟵ growth\n• biological_process ⟵ is_a ⟵ locomotion\n• biological_process ⟵ is_a ⟵ homeostatic process\n• biological_process ⟵ is_a ⟵ pigmentation\n• biological_process ⟵ is_a ⟵ biological process involved in interspecies interaction between organisms\n• biological_process ⟵ is_a ⟵ biological phase\n• biological_process ⟵ is_a ⟵ rhythmic process\n• biological_process ⟵ positively_regulates ⟵ positive regulation of biological process\n• biological_process ⟵ negatively_regulates ⟵ negative regulation of biological process\n• biological_process ⟵ regulates ⟵ regulation of biological process\n• biological_process ⟵ is_a ⟵ response to stimulus\n• biological_process ⟵ is_a ⟵ localization\n• biological_process ⟵ is_a ⟵ biological process involved in intraspecies interaction between organisms\n• biological_process ⟵ is_a ⟵ biological regulation\n• biological_process ⟵ is_a ⟵ detoxification\n**********************\n• molecular_function ⟵ is_a ⟵ cytoskeletal motor activity\n• molecular_function ⟵ is_a ⟵ catalytic activity\n• molecular_function ⟵ is_a ⟵ structural molecule activity\n• molecular_function ⟵ is_a ⟵ transporter activity\n• molecular_function ⟵ is_a ⟵ binding\n• molecular_function ⟵ is_a ⟵ electron transfer activity\n• molecular_function ⟵ is_a ⟵ antioxidant activity\n• molecular_function ⟵ is_a ⟵ cargo receptor activity\n• molecular_function ⟵ is_a ⟵ protein folding chaperone\n• molecular_function ⟵ is_a ⟵ translation regulator activity\n• molecular_function ⟵ is_a ⟵ nutrient reservoir activity\n• molecular_function ⟵ is_a ⟵ molecular transducer activity\n• molecular_function ⟵ is_a ⟵ molecular adaptor activity\n• molecular_function ⟵ is_a ⟵ toxin activity\n• molecular_function ⟵ is_a ⟵ molecular function regulator activity\n• molecular_function ⟵ is_a ⟵ dirigent protein activity\n• molecular_function ⟵ is_a ⟵ molecular carrier activity\n• molecular_function ⟵ is_a ⟵ transcription regulator activity\n• molecular_function ⟵ is_a ⟵ general transcription initiation factor activity\n• molecular_function ⟵ is_a ⟵ molecular sequestering activity\n• molecular_function ⟵ is_a ⟵ molecular template activity\n• molecular_function ⟵ is_a ⟵ fusogenic activity\n• molecular_function ⟵ is_a ⟵ ATP-dependent activity\n• molecular_function ⟵ is_a ⟵ RNA folding chaperone\n• molecular_function ⟵ is_a ⟵ protein-containing complex destabilizing activity\n• molecular_function ⟵ is_a ⟵ protein-containing complex stabilizing activity\n• molecular_function ⟵ is_a ⟵ pore-forming activity\n• molecular_function ⟵ is_a ⟵ membrane destabilizing activity\n• molecular_function ⟵ is_a ⟵ molecular tag activity\n• molecular_function ⟵ is_a ⟵ membrane bending activity\n• molecular_function ⟵ is_a ⟵ membrane grommet activity\n• molecular_function ⟵ is_a ⟵ translation factor activity\n","output_type":"stream"}],"execution_count":50},{"cell_type":"code","source":"# Find edges to children terms\nnode = name_to_id[\"cellular_component\"]\nfor child, parent, key in go_graph.in_edges(node, keys=True):\n    print(f\"{parent} ⟵ {key} ⟵ {child}\")\n    while child:\n        print(f\"{parent} ⟵ {key} ⟵ {child}\")","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T05:10:22.899840Z","iopub.execute_input":"2025-12-04T05:10:22.900189Z","iopub.status.idle":"2025-12-04T05:10:22.905816Z","shell.execute_reply.started":"2025-12-04T05:10:22.900166Z","shell.execute_reply":"2025-12-04T05:10:22.904896Z"}},"outputs":[{"name":"stdout","text":"GO:0005575 ⟵ is_a ⟵ GO:0032991\nGO:0005575 ⟵ is_a ⟵ GO:0044423\nGO:0005575 ⟵ is_a ⟵ GO:0110165\n","output_type":"stream"}],"execution_count":51},{"cell_type":"code","source":"go_graph.in_edges(node, keys=True)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T05:03:27.856273Z","iopub.execute_input":"2025-12-04T05:03:27.856624Z","iopub.status.idle":"2025-12-04T05:03:27.863288Z","shell.execute_reply.started":"2025-12-04T05:03:27.856598Z","shell.execute_reply":"2025-12-04T05:03:27.862238Z"}},"outputs":[{"execution_count":47,"output_type":"execute_result","data":{"text/plain":"InMultiEdgeDataView([('GO:0032991', 'GO:0005575', 'is_a'), ('GO:0044423', 'GO:0005575', 'is_a'), ('GO:0110165', 'GO:0005575', 'is_a')])"},"metadata":{}}],"execution_count":47},{"cell_type":"code","source":"go_graph.out_edges(node, keys=True)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-12-04T05:03:49.247334Z","iopub.execute_input":"2025-12-04T05:03:49.247683Z","iopub.status.idle":"2025-12-04T05:03:49.253717Z","shell.execute_reply.started":"2025-12-04T05:03:49.247657Z","shell.execute_reply":"2025-12-04T05:03:49.252812Z"}},"outputs":[{"execution_count":48,"output_type":"execute_result","data":{"text/plain":"OutMultiEdgeDataView([])"},"metadata":{}}],"execution_count":48},{"cell_type":"markdown","source":"## Introduce the ESM2 embeddings of dimension 320","metadata":{}},{"cell_type":"code","source":"emb_df = pl.read_parquet('/kaggle/input/cafa6-protein-go-terms-feat-labels/train_protein_features_esm2_320.parquet')\nemb_df.shape","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:41:58.658688Z","iopub.execute_input":"2025-11-29T13:41:58.659031Z","iopub.status.idle":"2025-11-29T13:42:01.729354Z","shell.execute_reply.started":"2025-11-29T13:41:58.659006Z","shell.execute_reply":"2025-11-29T13:42:01.728348Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"emb_df.head(3)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:38:10.064873Z","iopub.execute_input":"2025-11-29T03:38:10.065198Z","iopub.status.idle":"2025-11-29T03:38:10.093707Z","shell.execute_reply.started":"2025-11-29T03:38:10.065172Z","shell.execute_reply":"2025-11-29T03:38:10.092751Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"dd.sql(\" select count(distinct(protein_accession_id)) as proteins from emb_df \").pl()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:38:16.750007Z","iopub.execute_input":"2025-11-29T03:38:16.750939Z","iopub.status.idle":"2025-11-29T03:38:17.073070Z","shell.execute_reply.started":"2025-11-29T03:38:16.750905Z","shell.execute_reply":"2025-11-29T03:38:17.071893Z"}},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"## Bring in the training data; proteins and the corresponding GO terms","metadata":{}},{"cell_type":"code","source":"train_terms_df = pl.read_csv('/kaggle/input/cafa-6-protein-function-prediction/Train/train_terms.tsv', separator='\\t')\ntrain_terms_df.shape","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:42:10.846984Z","iopub.execute_input":"2025-11-29T13:42:10.847326Z","iopub.status.idle":"2025-11-29T13:42:11.398119Z","shell.execute_reply.started":"2025-11-29T13:42:10.847302Z","shell.execute_reply":"2025-11-29T13:42:11.397297Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"train_terms_df.head(3)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:38:53.132796Z","iopub.execute_input":"2025-11-29T03:38:53.133537Z","iopub.status.idle":"2025-11-29T03:38:53.140000Z","shell.execute_reply.started":"2025-11-29T03:38:53.133504Z","shell.execute_reply":"2025-11-29T03:38:53.139008Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"dd.sql(\" select count(distinct(EntryID)) as proteins from train_terms_df \").pl()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:39:02.199300Z","iopub.execute_input":"2025-11-29T03:39:02.199641Z","iopub.status.idle":"2025-11-29T03:39:02.260595Z","shell.execute_reply.started":"2025-11-29T03:39:02.199618Z","shell.execute_reply":"2025-11-29T03:39:02.259501Z"}},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"## Introduce the embeddings into the training data","metadata":{}},{"cell_type":"code","source":"train_terms_w_embeds = dd.sql(\"select distinct t1.protein_accession_id, t2.term, t2.aspect \\\n, t1.embedding_arrays as protein_embedding \\\nfrom emb_df t1 \\\njoin train_terms_df t2 \\\non t1.protein_accession_id = t2.EntryID\").pl()\n\ntrain_terms_w_embeds.shape","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:42:16.275464Z","iopub.execute_input":"2025-11-29T13:42:16.275776Z","iopub.status.idle":"2025-11-29T13:42:19.721381Z","shell.execute_reply.started":"2025-11-29T13:42:16.275754Z","shell.execute_reply":"2025-11-29T13:42:19.720514Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"train_terms_w_embeds.head(3)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:40:07.728476Z","iopub.execute_input":"2025-11-29T03:40:07.729354Z","iopub.status.idle":"2025-11-29T03:40:07.736370Z","shell.execute_reply.started":"2025-11-29T03:40:07.729322Z","shell.execute_reply":"2025-11-29T03:40:07.735446Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"dd.sql(\"select count(distinct(protein_accession_id)) as uniq_protein_accession_ids from train_terms_w_embeds\").pl()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:40:25.167894Z","iopub.execute_input":"2025-11-29T03:40:25.168245Z","iopub.status.idle":"2025-11-29T03:40:25.219198Z","shell.execute_reply.started":"2025-11-29T03:40:25.168222Z","shell.execute_reply":"2025-11-29T03:40:25.218299Z"}},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"## Create separate dataframes for each root ontology","metadata":{}},{"cell_type":"code","source":"train_terms_w_embeds_bp = train_terms_w_embeds.filter(pl.col(\"term\").is_in(separated_terms['BP']))\nprint(\"train_terms_w_embeds_bp shape -> \",train_terms_w_embeds_bp.shape)\n\ntrain_terms_w_embeds_cc = train_terms_w_embeds.filter(pl.col(\"term\").is_in(separated_terms['CC']))\nprint(\"train_terms_w_embeds_cc shape -> \",train_terms_w_embeds_cc.shape)\n\ntrain_terms_w_embeds_mf = train_terms_w_embeds.filter(pl.col(\"term\").is_in(separated_terms['MF']))\nprint(\"train_terms_w_embeds_mf shape -> \",train_terms_w_embeds_mf.shape)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:42:26.822168Z","iopub.execute_input":"2025-11-29T13:42:26.822492Z","iopub.status.idle":"2025-11-29T13:42:27.514690Z","shell.execute_reply.started":"2025-11-29T13:42:26.822472Z","shell.execute_reply":"2025-11-29T13:42:27.513767Z"}},"outputs":[],"execution_count":null},{"cell_type":"markdown","source":"## Get the 1500 most frequently occuring GO terms","metadata":{}},{"cell_type":"code","source":"plot_df = dd.sql(\"select term, count(1) as freq from train_terms_w_embeds_bp group by term order by count(1) desc limit 100\").pl()\nplot_df_pandas = plot_df.to_pandas()\n\nfigure, axis = plt.subplots(1, 1, figsize=(12, 6))\n\nbp = sns.barplot(ax=axis, x=plot_df_pandas.term, y=plot_df_pandas.freq)\nbp.set_xticklabels(bp.get_xticklabels(), rotation=90, size = 6)\naxis.set_title('Top 100 frequent GO term IDs')\nbp.set_xlabel(\"GO term IDs\", fontsize = 12)\nbp.set_ylabel(\"Count\", fontsize = 12)\nplt.show()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:47:55.726660Z","iopub.execute_input":"2025-11-29T03:47:55.727420Z","iopub.status.idle":"2025-11-29T03:47:56.718651Z","shell.execute_reply.started":"2025-11-29T03:47:55.727392Z","shell.execute_reply":"2025-11-29T03:47:56.717730Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"plot_df = dd.sql(\"select term, count(1) as freq from train_terms_w_embeds_cc group by term order by count(1) desc limit 100\").pl()\nplot_df_pandas = plot_df.to_pandas()\n\nfigure, axis = plt.subplots(1, 1, figsize=(12, 6))\n\nbp = sns.barplot(ax=axis, x=plot_df_pandas.term, y=plot_df_pandas.freq)\nbp.set_xticklabels(bp.get_xticklabels(), rotation=90, size = 6)\naxis.set_title('Top 100 frequent GO term IDs')\nbp.set_xlabel(\"GO term IDs\", fontsize = 12)\nbp.set_ylabel(\"Count\", fontsize = 12)\nplt.show()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:48:27.535910Z","iopub.execute_input":"2025-11-29T03:48:27.536238Z","iopub.status.idle":"2025-11-29T03:48:28.400285Z","shell.execute_reply.started":"2025-11-29T03:48:27.536217Z","shell.execute_reply":"2025-11-29T03:48:28.399424Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"plot_df = dd.sql(\"select term, count(1) as freq from train_terms_w_embeds_mf group by term order by count(1) desc limit 100\").pl()\nplot_df_pandas = plot_df.to_pandas()\n\nfigure, axis = plt.subplots(1, 1, figsize=(12, 6))\n\nbp = sns.barplot(ax=axis, x=plot_df_pandas.term, y=plot_df_pandas.freq)\nbp.set_xticklabels(bp.get_xticklabels(), rotation=90, size = 6)\naxis.set_title('Top 100 frequent GO term IDs')\nbp.set_xlabel(\"GO term IDs\", fontsize = 12)\nbp.set_ylabel(\"Count\", fontsize = 12)\nplt.show()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T03:48:56.554883Z","iopub.execute_input":"2025-11-29T03:48:56.555239Z","iopub.status.idle":"2025-11-29T03:48:57.416989Z","shell.execute_reply.started":"2025-11-29T03:48:56.555214Z","shell.execute_reply":"2025-11-29T03:48:57.416006Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"from typing import List\n\nnum_of_labels = 1500\n\ndef train_labels_per_ontology_pl(input_df: pl.DataFrame, input_num_of_labels: int) -> pl.DataFrame:\n    \"\"\"\n    Filters a Polars DataFrame to include only the top N most frequent terms.\n    \n    Args:\n        input_df: The input Polars DataFrame with a 'term' column.\n        input_num_of_labels: The number of top terms to keep.\n        \n    Returns:\n        A new Polars DataFrame containing only the top N frequent labels.\n    \"\"\"\n\n    # 1. Calculate the top N frequent terms using Polars methods\n    top_labels_df = input_df.group_by(\"term\").count().sort(\"count\", descending=True).limit(input_num_of_labels)\n    \n    # 2. Extract these top terms into a Python list\n    labels_list: List[str] = top_labels_df[\"term\"].to_list()\n    \n    # 3. Filter the original DataFrame to keep only those terms\n    train_labels = input_df.filter(pl.col(\"term\").is_in(labels_list))\n    \n    print(f\"Shape of filtered training data: {train_labels.shape}\")\n    print(\"Head of filtered training data:\")\n    print(train_labels.head())\n\n    return train_labels\n","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:42:39.994925Z","iopub.execute_input":"2025-11-29T13:42:39.995267Z","iopub.status.idle":"2025-11-29T13:42:40.001633Z","shell.execute_reply.started":"2025-11-29T13:42:39.995213Z","shell.execute_reply":"2025-11-29T13:42:40.000692Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"train_terms_bp = train_labels_per_ontology_pl(train_terms_w_embeds_bp, num_of_labels)\ntrain_terms_cc = train_labels_per_ontology_pl(train_terms_w_embeds_cc, num_of_labels)\ntrain_terms_mf = train_labels_per_ontology_pl(train_terms_w_embeds_mf, num_of_labels)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:42:49.287719Z","iopub.execute_input":"2025-11-29T13:42:49.288107Z","iopub.status.idle":"2025-11-29T13:42:50.529431Z","shell.execute_reply.started":"2025-11-29T13:42:49.288084Z","shell.execute_reply":"2025-11-29T13:42:50.528507Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"print(dd.sql(\"select count(distinct(protein_accession_id)) as uniq_protein_accession_ids from train_terms_bp\").pl())\nprint(dd.sql(\"select count(distinct(protein_accession_id)) as uniq_protein_accession_ids from train_terms_cc\").pl())\nprint(dd.sql(\"select count(distinct(protein_accession_id)) as uniq_protein_accession_ids from train_terms_mf\").pl())","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:42:58.546921Z","iopub.execute_input":"2025-11-29T13:42:58.547254Z","iopub.status.idle":"2025-11-29T13:42:58.635246Z","shell.execute_reply.started":"2025-11-29T13:42:58.547207Z","shell.execute_reply":"2025-11-29T13:42:58.634489Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"train_protein_labels_bp = train_terms_bp.select(['protein_accession_id','term'])\nprint(train_protein_labels_bp.head(5))\n\ntrain_protein_labels_cc = train_terms_cc.select(['protein_accession_id','term'])\nprint(train_protein_labels_cc.head(5))\n\ntrain_protein_labels_mf = train_terms_mf.select(['protein_accession_id','term'])\nprint(train_protein_labels_mf.head(5))","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:43:06.626199Z","iopub.execute_input":"2025-11-29T13:43:06.626607Z","iopub.status.idle":"2025-11-29T13:43:06.634757Z","shell.execute_reply.started":"2025-11-29T13:43:06.626573Z","shell.execute_reply":"2025-11-29T13:43:06.633902Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"def get_pivoted_df(input_df: pl.DataFrame) -> pl.DataFrame:\n    train_protein_labels_w_term_presence = input_df.with_columns(\n        pl.lit(1.0).alias(\"presence\")\n    )\n\n    pivoted_df = train_protein_labels_w_term_presence.pivot(\n        index=\"protein_accession_id\",\n        on=\"term\",\n        values=\"presence\",  \n        aggregate_function=\"sum\"\n    )\n    \n    pivoted_df = pivoted_df.fill_null(0)\n    print(pivoted_df.head(5))\n    return pivoted_df","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:43:12.365345Z","iopub.execute_input":"2025-11-29T13:43:12.365663Z","iopub.status.idle":"2025-11-29T13:43:12.371850Z","shell.execute_reply.started":"2025-11-29T13:43:12.365640Z","shell.execute_reply":"2025-11-29T13:43:12.370858Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"pivoted_df_bp = get_pivoted_df(train_protein_labels_bp)\npivoted_df_cc = get_pivoted_df(train_protein_labels_cc)\npivoted_df_mf = get_pivoted_df(train_protein_labels_mf)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:43:16.619798Z","iopub.execute_input":"2025-11-29T13:43:16.620781Z","iopub.status.idle":"2025-11-29T13:43:21.972088Z","shell.execute_reply.started":"2025-11-29T13:43:16.620754Z","shell.execute_reply":"2025-11-29T13:43:21.970929Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"print(pivoted_df_bp.shape)\nprint(pivoted_df_cc.shape)\nprint(pivoted_df_mf.shape)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:43:30.334086Z","iopub.execute_input":"2025-11-29T13:43:30.334503Z","iopub.status.idle":"2025-11-29T13:43:30.340071Z","shell.execute_reply.started":"2025-11-29T13:43:30.334478Z","shell.execute_reply":"2025-11-29T13:43:30.338784Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"train_protein_features_bp = dd.sql(\"select distinct protein_accession_id, protein_embedding from train_terms_bp\").pl()\nprint(train_protein_features_bp.head(5))\n\ntrain_protein_features_cc = dd.sql(\"select distinct protein_accession_id, protein_embedding from train_terms_cc\").pl()\nprint(train_protein_features_cc.head(5))\n\ntrain_protein_features_mf = dd.sql(\"select distinct protein_accession_id, protein_embedding from train_terms_mf\").pl()\nprint(train_protein_features_mf.head(5))","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:43:42.186669Z","iopub.execute_input":"2025-11-29T13:43:42.186990Z","iopub.status.idle":"2025-11-29T13:43:43.810858Z","shell.execute_reply.started":"2025-11-29T13:43:42.186968Z","shell.execute_reply":"2025-11-29T13:43:43.809923Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"pivoted_df_mf.write_parquet('train_protein_labels_mf.parquet')\ntrain_protein_features_mf.write_parquet('train_protein_features_mf.parquet')\n\npivoted_df_cc.write_parquet('train_protein_labels_cc.parquet')\ntrain_protein_features_cc.write_parquet('train_protein_features_cc.parquet')\n\npivoted_df_bp.write_parquet('train_protein_labels_bp.parquet')\ntrain_protein_features_bp.write_parquet('train_protein_features_bp.parquet')","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:50:27.574130Z","iopub.execute_input":"2025-11-29T13:50:27.574775Z","iopub.status.idle":"2025-11-29T13:50:32.747588Z","shell.execute_reply.started":"2025-11-29T13:50:27.574749Z","shell.execute_reply":"2025-11-29T13:50:32.746620Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"test_prot_ids = [seq_record.id \n                 for seq_record in \n                 SeqIO.parse(\"/kaggle/input/cafa-6-protein-function-prediction/Test/testsuperset.fasta\", \"fasta\")]","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:54:34.854538Z","iopub.execute_input":"2025-11-29T13:54:34.854839Z","iopub.status.idle":"2025-11-29T13:54:36.690478Z","shell.execute_reply.started":"2025-11-29T13:54:34.854820Z","shell.execute_reply":"2025-11-29T13:54:36.689622Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"test_emb_df = pl.read_parquet('/kaggle/input/cafa6-protein-go-terms-feat-labels/test_protein_features_esm2_320.parquet')\ntest_emb_df.shape","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:53:53.834025Z","iopub.execute_input":"2025-11-29T13:53:53.834366Z","iopub.status.idle":"2025-11-29T13:54:00.687790Z","shell.execute_reply.started":"2025-11-29T13:53:53.834343Z","shell.execute_reply":"2025-11-29T13:54:00.686792Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"test_embeds = test_emb_df.filter(pl.col(\"protein_accession_id\").is_in(test_prot_ids))\ntest_embeds.shape","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:54:51.265847Z","iopub.execute_input":"2025-11-29T13:54:51.266159Z","iopub.status.idle":"2025-11-29T13:54:51.330762Z","shell.execute_reply.started":"2025-11-29T13:54:51.266139Z","shell.execute_reply":"2025-11-29T13:54:51.329922Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"test_emb_df.head(4)","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:54:17.251798Z","iopub.execute_input":"2025-11-29T13:54:17.252160Z","iopub.status.idle":"2025-11-29T13:54:17.260338Z","shell.execute_reply.started":"2025-11-29T13:54:17.252137Z","shell.execute_reply":"2025-11-29T13:54:17.259320Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"dd.sql(\"select count(distinct(protein_accession_id)) as uniq_protein_accession_ids from test_embeds\").pl()","metadata":{"trusted":true,"execution":{"iopub.status.busy":"2025-11-29T13:55:03.002505Z","iopub.execute_input":"2025-11-29T13:55:03.002837Z","iopub.status.idle":"2025-11-29T13:55:03.042035Z","shell.execute_reply.started":"2025-11-29T13:55:03.002814Z","shell.execute_reply":"2025-11-29T13:55:03.041040Z"}},"outputs":[],"execution_count":null},{"cell_type":"code","source":"","metadata":{"trusted":true},"outputs":[],"execution_count":null}]}